Docs to split sharded database (#3004)
dimagi/commcare-cloud: docs/services/postgresql/split-partitioned-database.md

        'PORT': '5432',
        'TEST': {
            'SERIALIZE': False,
        },
    },
    'p1': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'commcarehq_p1',
        'USER': 'commcarehq',
        'PASSWORD': 'commcarehq',
        'HOST': 'pg1',
        'PORT': '5432',
        'TEST': {
            'SERIALIZE': False,
        },
    },
}

At the end of this process shards 0 & 1 should stay on p1 and shards 2 & 3 will be on p2.
```

## Process Overview

1. Setup *pg2* node as a standby node of *pg1*
2. Promote *pg2* to a master node
3. Update the configuration so that requests for database alias *p2* go to *pg2* instead
of *pg1*.
4. Delete invalid data

## Process detail

### 1. Setup *pg2* as a standby node of *pg1*
This step does not require downtime and can be done at any stage prior to the
downtime.

```bash
commcare-cloud <env> ansible-playbook setup_pg_standby.yml -e standby=pg2 -e hot_standby_master=pg1 -e replication_slot=[replication slot name]
```

### 2. Stop all DB requests
This will bring the CommCare HQ site down.

**Stop all CommCare processes**
```bash
commcare-cloud <env> downtime start
```

**Stop pgbouncer**

```bash

> commcare-cloud <env> run-shell-command 'monit stop pgbouncer' --become

Reviewer: or `commcare-cloud <env> service postgres stop --only pgbouncer`

```

### 3. Check document counts in the databases
This step is useful for verifying the move at the end.
```
commcare-cloud <env> django-manage print_approximate_doc_distribution --csv
```

### 4. Update database name

In pg2 postgres shell:

```sql

> ALTER DATABASE commcarehq_p1 RENAME TO commcarehq_p2;

Author: ALso needs to update pgbouncer config after this

# Splitting a shard in postgresql

> 

Author: Should add introduction

        'NAME': 'commcarehq_proxy',
        'USER': 'commcarehq',
        'PASSWORD': 'commcarehq',
        'HOST': 'pg1',
        'PORT': '5432',
        'TEST': {
            'SERIALIZE': False,
        },
    },
    'p1': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'commcarehq_p1',
        'USER': 'commcarehq',
        'PASSWORD': 'commcarehq',
        'HOST': 'pg1',
        'PORT': '5432',
        'TEST': {
            'SERIALIZE': False,
        },
    },
}
```

## Process Overview

1. Setup *pg2* node as a standby node of *pg1*
2. Promote *pg2* to a master node
3. Update the configuration so that requests for database alias *p2* go to *pg2* instead
of *pg1*.
4. Delete invalid data

## Process detail

### 1. Setup *pg2* as a standby node of *pg1*
This step does not require downtime and can be done at any stage prior to the
downtime.

```bash
commcare-cloud <env> ansible-playbook setup_pg_standby.yml -e standby=pg2 -e hot_standby_master=pg1 -e replication_slot=[replication slot name]
```

### 2. Stop all DB requests
This will bring the CommCare HQ site down.

**Stop all CommCare processes**
```bash
commcare-cloud <env> downtime start
```

> **Stop pgbouncer**

Author: I think there is probably a better commcare-cloud command for this

To be completely certain that no data will be updating during the move you can also
prevent connections from pgbouncer:

```bash
pg1 $ psql -p 6432 -U someuser pgbouncer

> PAUSE commcarehq_p1
```

### 3. Check document counts in the databases
This step is useful for verifying the move at the end.
```
commcare-cloud <env> django-manage print_approximate_doc_distribution --csv
```

### 4. Update database name

In pg2 postgres shell:

```sql
ALTER DATABASE commcarehq_p1 RENAME TO commcarehq_p2;

> ```

Author: There's currently no way to manually specify a database name for a sharded database. It must be `commcarehq_django_alias`.

Reviewer: sure but the name comes from the Django DB settings right? I mean I would still want to rename the DB.

### 5. Update configuration

**Update ansible config**

Update the *dbs* variable in the environment's *postgresql.yml* file
to show that the *p2* database is now on *pg2*:

```diff
...
 dbs:
 ...
   form_processing:
     ...
     partitions:
       p1:
-        shards: [0, 3]
+        shards: [0, 1]
         host: pg1
+      p2:
+        shards: [2, 3]
+        host: pg2
       ...
```

**Deploy changes**
```
# update localsettings
commcare-cloud <env> update-config

# update PostgreSQL config on new PG node

> commcare-cloud <env> ap deploy_db.yml --limit=pg2

Author: perhaps this should be both pg1 & pg2

```

### 6. Verify config changes
```
commcare-cloud <env> django-manage print_approximate_doc_distribution --csv
```

This should show that the *p2* database is now on the *pg2* host.

### 7. Promote *pg2* to master

**Verify that replication is up to date**
```
commcare-cloud <env> run-shell-command pg1,pg2 'ps -ef | grep -E "sender|receiver"'

    [ pg1 ] ps -ef | grep -E "sender|receiver"
    postgres 5295 4517 0 Jul24 ? 00:00:01 postgres: wal sender process rep 10.116.175.107(49770) streaming 0/205B598

    [ pg2 ] ps -ef | grep -E "sender|receiver"
    postgres 3821 3808 0 Jul24 ? 00:01:27 postgres: wal receiver process streaming 0/205B598
```

Output shows that master and standby are up to date (both processing the same log).

**Promote *pg2***
```bash
commcare-cloud <env> run-shell-command pg2 -b 'pg_ctlcluster <pg version> main promote'
```

### 8. Verify doc counts

> Re-run command from step 5 to verify that the document counts are double the number before.

Author: ```suggestion
Re-run command from step 6 to verify that the document counts are double the number before.
```

This is because each shard's data is now in two databases.

### 9. Update pl_proxy config
```
commcare-cloud <env> django-manage configure_pl_proxy_cluster
```

### 10. Remove invalid data
The same data now exists in multiple shards. Run the following command to delete that data

```
commcare-cloud <env> django-manage locate_invalid_shard_data --delete
```

Optionally you can run a vacuum to more reclaim space from the table.

```sql
VACUUM FULL table_name;
```

### 11. Verify doc counts

> Re-run command from step 5 to verify that the document counts are the same number as before.

Author: ```suggestion
Re-run command from step 6 to verify that the document counts are the same number as in step 5 (half as in step 8).
```

### 12. Restart services
**Unpause pgbouncer**
```bash
pg1 $ psql -p 6543 -U someuser pgbouncer

> RESUME commcarehq_p1
```

**Restart services**
```bash

> commcare-cloud <env> fab supervisorctl:"start all"

Author: ```suggestion
commcare-cloud <env> downtime end
```

```

> ### 13. Validate the setup

Author: should likely delete this?
